File: parse_ged.js
const fs = require('fs');

// Read GEDCOM file, handle CR-only line endings from MacFamilyTree
const raw = fs.readFileSync('My.ged', 'utf-8');
const lines = raw.replace(/\r\n/g, '\n').replace(/\r/g, '\n').split('\n');

const individuals = {};
const families = {};
const notes = {};
const objes = {};

let currentRecord = null;
let currentType = null;
let currentSubTag = null;

for (let i = 0; i < lines.length; i++) {
  const line = lines[i];
  const match = line.match(/^(\d+)\s+(.*)$/);
  if (!match) continue;

  const level = parseInt(match[1]);
  const rest = match[2];

  if (level === 0) {
    currentSubTag = null;
    // Check for record start: 0 @ID@ TYPE
    const recMatch = rest.match(/^@(\d+)@\s+(\w+)$/);
    if (recMatch) {
      const id = recMatch[1];
      const type = recMatch[2];
      currentType = type;
      if (type === 'INDI') {
        currentRecord = {
          id,
          givn: '',
          surn: '',
          secg: '',
          sex: '',
          birthDate: '',
          birthPlace: '',
          deathDate: '',
          deathPlace: '',
          occupation: '',
          residence: '',
          objes: [],
          fams: [],
          famc: [],
          notes: [],
          photoIds: []
        };
        individuals[id] = currentRecord;
      } else if (type === 'FAM') {
        currentRecord = {
          id,
          husb: '',
          wife: '',
          chil: [],
          marriageDate: '',
          marriagePlace: ''
        };
        families[id] = currentRecord;
      } else if (type === 'NOTE') {
        currentRecord = { id, text: '' };
        notes[id] = currentRecord;
      } else if (type === 'OBJE') {
        currentRecord = { id, file: '', form: '', title: '' };
        objes[id] = currentRecord;
      } else {
        currentRecord = null;
        currentType = null;
      }
    } else {
      currentRecord = null;
      currentType = null;
    }
    continue;
  }

  if (!currentRecord) continue;

  const tagMatch = rest.match(/^(\w+)\s*(.*)$/);
  if (!tagMatch) continue;

  const tag = tagMatch[1];
  const value = tagMatch[2].trim();

  if (currentType === 'INDI') {
    if (level === 1) {
      currentSubTag = tag;
      if (tag === 'GIVN') currentRecord.givn = value;
      else if (tag === 'SURN') currentRecord.surn = value;
      else if (tag === 'SECG') currentRecord.secg = value;
      else if (tag === 'SEX') currentRecord.sex = value;
      else if (tag === 'BIRT') currentSubTag = 'BIRT';
      else if (tag === 'DEAT') currentSubTag = 'DEAT';
      else if (tag === 'OCCU') currentRecord.occupation = value;
      else if (tag === 'RESI') currentSubTag = 'RESI';
      else if (tag === 'OBJE') {
        const objeRef = value.match(/@(\d+)@/);
        if (objeRef) currentRecord.objes.push(objeRef[1]);
      }
      else if (tag === 'FAMS') {
        const ref = value.match(/@(\d+)@/);
        if (ref) currentRecord.fams.push(ref[1]);
      }
      else if (tag === 'FAMC') {
        const ref = value.match(/@(\d+)@/);
        if (ref) currentRecord.famc.push(ref[1]);
      }
      else if (tag === 'NOTE') {
        const ref = value.match(/@(\d+)@/);
        if (ref) currentRecord.notes.push(ref[1]);
      }
      else if (tag === 'NAME') {
        // Only parse the first NAME (level 1) for givn/surn if not yet set
        // NAME format: "Given /Surname/"
        if (!currentRecord._nameProcessed) {
          const nameMatch = value.match(/^(.*?)\s*\/(.*?)\//);
          if (nameMatch) {
            if (!currentRecord.givn) currentRecord.givn = nameMatch[1].trim();
            if (!currentRecord.surn) currentRecord.surn = nameMatch[2].trim();
          }
          currentRecord._nameProcessed = true;
        }
      }
    } else if (level === 2) {
      if (tag === 'GIVN') currentRecord.givn = value;
      else if (tag === 'SURN') currentRecord.surn = value;
      else if (tag === 'SECG') currentRecord.secg = value;
      else if (currentSubTag === 'BIRT') {
        if (tag === 'DATE') currentRecord.birthDate = value;
        else if (tag === 'PLAC') currentRecord.birthPlace = value;
        else if (tag === 'ADDR') currentRecord.birthPlace = value;
      } else if (currentSubTag === 'DEAT') {
        if (tag === 'DATE') currentRecord.deathDate = value;
        else if (tag === 'PLAC') currentRecord.deathPlace = value;
      } else if (currentSubTag === 'RESI') {
        if (tag === 'ADDR') currentRecord.residence = value;
      }
    }
  } else if (currentType === 'FAM') {
    if (level === 1) {
      currentSubTag = tag;
      if (tag === 'HUSB') {
        const ref = value.match(/@(\d+)@/);
        if (ref) currentRecord.husb = ref[1];
      } else if (tag === 'WIFE') {
        const ref = value.match(/@(\d+)@/);
        if (ref) currentRecord.wife = ref[1];
      } else if (tag === 'CHIL') {
        const ref = value.match(/@(\d+)@/);
        if (ref) currentRecord.chil.push(ref[1]);
      } else if (tag === 'MARR') {
        currentSubTag = 'MARR';
      }
    } else if (level === 2) {
      if (currentSubTag === 'MARR') {
        if (tag === 'DATE') currentRecord.marriageDate = value;
        else if (tag === 'PLAC') currentRecord.marriagePlace = value;
      }
    }
  } else if (currentType === 'NOTE') {
    if (tag === 'CONT' || tag === 'CONC') {
      currentRecord.text += (currentRecord.text ? '\n' : '') + value;
    }
  } else if (currentType === 'OBJE') {
    if (level === 1) {
      if (tag === 'FILE') currentRecord.file = value;
      else if (tag === 'TITL') currentRecord.title = value;
    } else if (level === 2) {
      if (tag === 'FORM') currentRecord.form = value;
    }
  }
}

// Build photoIds for each individual from OBJE references
for (const indi of Object.values(individuals)) {
  indi.photoIds = [];
  for (const objeId of indi.objes) {
    const obje = objes[objeId];
    if (obje && obje.file) {
      // Extract the numeric ID from filename like "54203864.jpg"
      const fileId = obje.file.replace(/\.\w+$/, '');
      indi.photoIds.push(fileId);
    }
  }
  delete indi._nameProcessed;
}

const data = { individuals, families, notes };

fs.writeFileSync('data.json', JSON.stringify(data, null, 2), 'utf-8');

console.log(`Parsed:`);
console.log(`  Individuals: ${Object.keys(individuals).length}`);
console.log(`  Families: ${Object.keys(families).length}`);
console.log(`  Notes: ${Object.keys(notes).length}`);
console.log(`  Media objects: ${Object.keys(objes).length}`);

// Check photo coverage
let withPhotos = 0;
for (const indi of Object.values(individuals)) {
  if (indi.photoIds.length > 0) withPhotos++;
}
console.log(`  Individuals with photos: ${withPhotos}`);

File: parse_ged.py
import json
import re

with open('My.ged', 'r', encoding='utf-8') as f:
    raw = f.read()

lines = raw.replace('\r\n', '\n').replace('\r', '\n').split('\n')

individuals = {}
families = {}
notes = {}
objes = {}

current_record = None
current_type = None
# Track level-1 and level-2 context tags
l1_tag = None
l1_is_first_name = False

for line in lines:
    m = re.match(r'^(\d+)\s+(.*)$', line)
    if not m:
        continue
    level = int(m.group(1))
    rest = m.group(2)

    if level == 0:
        l1_tag = None
        rec_m = re.match(r'^@(\d+)@\s+(\w+)$', rest)
        if rec_m:
            rid = rec_m.group(1)
            rtype = rec_m.group(2)
            current_type = rtype
            if rtype == 'INDI':
                current_record = {
                    'id': rid, 'givn': '', 'surn': '', 'secg': '', 'sex': '',
                    'birthDate': '', 'birthPlace': '', 'deathDate': '', 'deathPlace': '',
                    'occupation': '', 'residence': '',
                    'objes': [], 'fams': [], 'famc': [], 'notes': [], 'photoIds': [],
                    '_nameCount': 0
                }
                individuals[rid] = current_record
            elif rtype == 'FAM':
                current_record = {
                    'id': rid, 'husb': '', 'wife': '', 'chil': [],
                    'marriageDate': '', 'marriagePlace': ''
                }
                families[rid] = current_record
            elif rtype == 'NOTE':
                current_record = {'id': rid, 'text': ''}
                notes[rid] = current_record
            elif rtype == 'OBJE':
                current_record = {'id': rid, 'file': '', 'form': '', 'title': ''}
                objes[rid] = current_record
            else:
                current_record = None
                current_type = None
        else:
            current_record = None
            current_type = None
        continue

    if current_record is None:
        continue

    tag_m = re.match(r'^(\w+)\s*(.*)', rest)
    if not tag_m:
        continue
    tag = tag_m.group(1)
    value = tag_m.group(2).strip()

    if current_type == 'INDI':
        if level == 1:
            l1_tag = tag
            if tag == 'NAME':
                current_record['_nameCount'] += 1
                l1_is_first_name = (current_record['_nameCount'] == 1)
                # Parse inline name: "Given Middle /Surname/"
                if l1_is_first_name:
                    nm = re.match(r'^(.*?)\s*/(.+?)/', value)
                    if nm:
                        if not current_record['givn']:
                            current_record['givn'] = nm.group(1).strip()
                        if not current_record['surn']:
                            current_record['surn'] = nm.group(2).strip()
            elif tag == 'SEX':
                current_record['sex'] = value
            elif tag == 'BIRT':
                pass  # handled at level 2
            elif tag == 'DEAT':
                pass
            elif tag == 'OCCU':
                current_record['occupation'] = value
            elif tag == 'RESI':
                pass
            elif tag == 'OBJE':
                ref = re.search(r'@(\d+)@', value)
                if ref:
                    current_record['objes'].append(ref.group(1))
            elif tag == 'FAMS':
                ref = re.search(r'@(\d+)@', value)
                if ref:
                    current_record['fams'].append(ref.group(1))
            elif tag == 'FAMC':
                ref = re.search(r'@(\d+)@', value)
                if ref:
                    current_record['famc'].append(ref.group(1))
            elif tag == 'NOTE':
                ref = re.search(r'@(\d+)@', value)
                if ref:
                    current_record['notes'].append(ref.group(1))
        elif level == 2:
            if l1_tag == 'NAME' and l1_is_first_name:
                if tag == 'GIVN':
                    current_record['givn'] = value
                elif tag == 'SURN':
                    current_record['surn'] = value
                elif tag == 'SECG':
                    current_record['secg'] = value
            elif l1_tag == 'BIRT':
                if tag == 'DATE':
                    current_record['birthDate'] = value
                elif tag == 'PLAC':
                    current_record['birthPlace'] = value
                elif tag == 'ADDR':
                    if not current_record['birthPlace']:
                        current_record['birthPlace'] = value
            elif l1_tag == 'DEAT':
                if tag == 'DATE':
                    current_record['deathDate'] = value
                elif tag == 'PLAC':
                    current_record['deathPlace'] = value
            elif l1_tag == 'RESI':
                if tag == 'ADDR':
                    current_record['residence'] = value

    elif current_type == 'FAM':
        if level == 1:
            l1_tag = tag
            if tag == 'HUSB':
                ref = re.search(r'@(\d+)@', value)
                if ref:
                    current_record['husb'] = ref.group(1)
            elif tag == 'WIFE':
                ref = re.search(r'@(\d+)@', value)
                if ref:
                    current_record['wife'] = ref.group(1)
            elif tag == 'CHIL':
                ref = re.search(r'@(\d+)@', value)
                if ref:
                    current_record['chil'].append(ref.group(1))
        elif level == 2:
            if l1_tag == 'MARR':
                if tag == 'DATE':
                    current_record['marriageDate'] = value
                elif tag == 'PLAC':
                    current_record['marriagePlace'] = value

    elif current_type == 'NOTE':
        if tag in ('CONT', 'CONC'):
            if current_record['text']:
                current_record['text'] += '\n' + value
            else:
                current_record['text'] = value

    elif current_type == 'OBJE':
        if level == 1:
            if tag == 'FILE':
                current_record['file'] = value
            elif tag == 'TITL':
                current_record['title'] = value
        elif level == 2:
            if tag == 'FORM':
                current_record['form'] = value

# Build photoIds
for indi in individuals.values():
    indi['photoIds'] = []
    for obje_id in indi['objes']:
        obje = objes.get(obje_id)
        if obje and obje['file']:
            file_id = re.sub(r'\.\w+$', '', obje['file'])
            indi['photoIds'].append(file_id)
    del indi['_nameCount']

data = {'individuals': individuals, 'families': families, 'notes': notes}

with open('data.json', 'w', encoding='utf-8') as f:
    json.dump(data, f, ensure_ascii=False, indent=2)

print(f"Parsed:")
print(f"  Individuals: {len(individuals)}")
print(f"  Families: {len(families)}")
print(f"  Notes: {len(notes)}")
print(f"  Media objects: {len(objes)}")

with_photos = sum(1 for i in individuals.values() if i['photoIds'])
print(f"  Individuals with photos: {with_photos}")
